marlowe/src/Language/Marlowe/Core/V1/Semantics/Next.hs

```diff
@@ -0,0 +1,175 @@
+-----------------------------------------------------------------------------
+--
+-- Module      :  $Headers
+-- License     :  Apache 2.0
+--
+-- Stability   :  Experimental
+-- Portability :  Portable
+--
+-- | Types for Marlowe semantics
+--
+-----------------------------------------------------------------------------
+
+
+{-# LANGUAGE BlockArguments #-}
+{-# LANGUAGE DataKinds #-}
+
+{-# LANGUAGE DeriveAnyClass #-}
+{-# LANGUAGE DeriveGeneric #-}
+{-# LANGUAGE DerivingVia #-}
+{-# LANGUAGE FlexibleContexts #-}
+{-# LANGUAGE FlexibleInstances #-}
+
+{-# LANGUAGE LambdaCase #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+
+{-# LANGUAGE NoImplicitPrelude #-}
+{-# LANGUAGE OverloadedStrings #-}
+{-# LANGUAGE RankNTypes #-}
+
+
+module Language.Marlowe.Core.V1.Semantics.Next
+  ( AmbiguousIntervalProvided(..)
+  , CaseIndex(..)
+  , Next(..)
+  , NextGeneralizedInput(..)
+  , next
+  ) where
+
+import Control.Applicative
+import Control.Monad.Cont ((=<<))
+import Data.Aeson
+import Data.Aeson.Types ()
+import Data.Bifunctor (Bifunctor(first))
+import Data.List.Index (indexed)
+import Data.List.NonEmpty as NEL (NonEmpty(..), nonEmpty, toList)
+import Data.Maybe (catMaybes)
+import Data.Text (pack)
+import Deriving.Aeson (Generic)
+import Language.Marlowe.Core.V1.Semantics
+import Language.Marlowe.Core.V1.Semantics.Types hiding (environment)
+import Language.Marlowe.Pretty (Pretty(..))
+import PlutusTx.Prelude (Bool(..), Either(..), Integer, Maybe(..), maybe, uncurry, ($), (.))
+import qualified Prelude as Haskell
+
+
+
+
+data Next = CanAdvance [NextGeneralizedInput]
+          | CanNotAdvance
+    deriving stock (Haskell.Show)
+
+data AmbiguousIntervalProvided = AmbiguousIntervalProvided
+
+newtype CaseIndex = CaseIndex {unIndex ::Integer}
+    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+    deriving anyclass (Pretty)
+    deriving (FromJSON,ToJSON)
+
+-- | Next Ranges of Inputs that can be applied to a given contract
+data NextGeneralizedInput
+  = CanDeposit CaseIndex Party AccountId Token Integer -- 1 possibility > 1 action
+  | CanChoose  CaseIndex ChoiceId [Bound] -- 1 possibility > many different actions
+  | CanNotify  CaseIndex -- 1 possibility > 1 action
+  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
+  deriving anyclass (Pretty)
```

Comment:
I want to suggest a couple of changes to this schema:

1. It is possible for a contract to be advanced or have inputs applied, so these should not be mutually exclusive cases.
2. It would be more useful if the input types were separated rather than combined in a sum type
   - It would be an error for more than one `CanNotify` to appear in the list
   - It makes it easier to ask questions like "can I make a deposit" or "can I make a choice" / "what deposits are available?", "what choices are available?"

```suggestion
data Next = Next
  { nextCanAdvance :: CanAdvance
  , nextCanNotify :: CanNotify
  , nextCanDeposit :: [CanDeposit]
  , nextCanChoose :: [CanChoose]
  }
  deriving stock (Haskell.Show, Haskell.Ord, Haskell.Eq, Generic)
  
data CanAdvance
  = CannotAdvance
  | CanAdvance
  deriving stock (Haskell.Show, Haskell.Ord, Haskell.Eq, Generic)
  deriving anyclass (Pretty)

newtype CaseIndex = CaseIndex {unIndex ::Integer}
    deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
    deriving anyclass (Pretty)
    deriving (FromJSON,ToJSON)

data CanNotify
  = CannotNotify
  | CanNotify CaseIndex (Maybe BuiltinByteString) -- Hash of the required merkleized continuation, if applicable
  deriving stock (Haskell.Show, Haskell.Ord, Haskell.Eq, Generic)
  deriving anyclass (Pretty)
  
data CanDeposist = CanDeposit CaseIndex Party AccountId Token Integer (Maybe BuiltinByteString)
  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
  deriving anyclass (Pretty)

data CanChoose = CanChoose CaseIndex ChoiceId [Bound] (Maybe BuiltinByteString)
  deriving stock (Haskell.Show,Haskell.Eq,Haskell.Ord,Generic)
  deriving anyclass (Pretty)

data AmbiguousIntervalProvided = AmbiguousIntervalProvided
```

Reply:
The Advance bit has another issue, it is not possible to tell whether empty inputs is allowed or not (that is what I was understanding by "CanAdvance"). You can find this out from the result of `reduceContractUntilQuiescent`, the first parameter is true if "empty inputs" is allowed, and false if not.

But that is independent of the rest of potential inputs, you can have the four cases:
- Empty inputs allowed, no other inputs allowed
- Empty inputs allowed, other inputs allowed
- Empty inputs disallowed, no other inputs allowed
- Empty inputs disallowed, other inputs allowed
These definitions are only considering 3 of them, I think they are missing the last one